### PR TITLE
[amp-sidebar 1.0][Toolbar] Added Body Spacing for Default Toolbar

### LINKED ISSUE #9985: [amp-sidebar 1.0][Toolbar] Create Manual tests
Evaluate Scenarios to be supported and create manual tests for all possible scenarios

- [ ] Sidebar with no toolbar
- [ ] Sidebar with one toolbar - min media, min & max media, max media, min 0px, no media (this scenario should not build the toolbar.)
- [ ] Sidebar with multiple toolbars on different media
- [ ] Sidebar with Multiple toolbars on same media (users could position on top or bottom)
- [ ] Sidebar with Multiple toolbars on same media and same layout (This will look ugly you should discuss with UX/Product on how to handle this)
- [ ] Sidebar with only toolbar content - needs ux / product decisions - we don't want to show empty sidebar if the sidebar only has toolbar contents

### PR DESCRIPTION
**Pull request by torch2424**
Closes #9985 
- Creates placeholder elements to push the body down
- Placeholder elements are recalculated on layout resize
- Made a better (AMP Start) Test manual file for toolbar, satisfying all manual test requirements.

Please see https://github.com/torch2424/amphtml/pull/8
for examples

### CODE REVIEW

**Comment by camelburrito** [comment]
Gave @torch2424 in-person feedback on this

**Comment by torch2424** [comment]
Please note in my latest changes, I created a new attribute called toolbar-container for when we are creating the container/target for the toolbar. This will allow our users to target the toolbar independintly with CSS. I was unable to create the new preety view for toolbar without this. And it would be required for #10126 , in order to apply styles (such as position:fixed) to only our generated containers.

File: extensions/amp-sidebar/1.0/amp-sidebar.css [hunk 1]
```diff
@@ -53,10 +53,10 @@
    z-index: 2147483646;
    width: 100%;
    position: fixed;
-   top: 0;
+   background-color: white;
  }
 
- .i-amphtml-toolbar > ul {
+ .i-amphtml-toolbar > ul:first-child {
```
Comment (camelburrito) [inline extensions/amp-sidebar/1.0/amp-sidebar.css]: shouldnt toolbar only have one nav element inside of it?

File: extensions/amp-sidebar/1.0/amp-sidebar.css [hunk 2]
```diff
@@ -53,10 +53,10 @@
    z-index: 2147483646;
    width: 100%;
    position: fixed;
-   top: 0;
+   background-color: white;
```
Comment (camelburrito) [inline extensions/amp-sidebar/1.0/amp-sidebar.css]: inherit?

File: extensions/amp-sidebar/1.0/amp-sidebar.css [hunk 3]
```diff
@@ -53,10 +53,10 @@
    z-index: 2147483646;
    width: 100%;
    position: fixed;
-   top: 0;
```
Comment (camelburrito) [inline extensions/amp-sidebar/1.0/amp-sidebar.css]: Why is this deleted?

File: extensions/amp-sidebar/1.0/amp-sidebar.js [hunk 4]
```diff
@@ -110,6 +110,7 @@ export class AmpSidebar extends AMP.BaseElement {
     const toolbarElements =
     Array.prototype.slice
       .call(this.element.querySelectorAll('nav[toolbar]'), 0);
+    toolbarElements.reverse();
```
Comment (camelburrito) [inline extensions/amp-sidebar/1.0/amp-sidebar.js]: why?

Comment (camelburrito) [inline extensions/amp-sidebar/1.0/amp-sidebar.js]: you should not do any DOM re-ordering

File: extensions/amp-sidebar/1.0/toolbar.js [hunk 5]
```diff
@@ -47,6 +50,9 @@ export class Toolbar {
     /** @private {Element|undefined} */
     this.targetElement_ = undefined;
 
+    /** @private {Element|undefined} */
+    this.placeholder_ = undefined;
```
Comment (camelburrito) [inline extensions/amp-sidebar/1.0/toolbar.js]: what is a placeholder here ? there is already a set definition for a placeholder in amp.

File: extensions/amp-sidebar/1.0/toolbar.js [hunk 6]
```diff
@@ -88,6 +94,15 @@ export class Toolbar {
     } else {
       this.hideToolbar_();
     }
+
+    // Use the placeholder to fill the height of the toolbar
+    this.vsync_.mutate(() => {
+      if (this.placeholder_) {
```
Comment (camelburrito) [inline extensions/amp-sidebar/1.0/toolbar.js]: This should be done in attempt show , you dont want to display the compensatingElement when toolbar is hidden. 

1. Check if there is a compensation element , if not create one and assign the height
2. else just assign the height
3. When you close the toolbar this has to be hidden

File: extensions/amp-sidebar/1.0/amp-sidebar.css (hunk 1 above)
Comment (torch2424) [inline extensions/amp-sidebar/1.0/amp-sidebar.css]: `i-amphtml-toolbar` is the class applied to the `<nav toolbar>` element. I noticed when implementing the sidebarOnly amp start, that these styles we being applied to the ul within the amp-accordian. So I ensured that this was the first child, as <nav toolbar> may only contain one direct child of `<ul>`

**Comment by camelburrito** [comment]
Lets not combine features and manual test files in the same PR. Please split it

File: extensions/amp-sidebar/1.0/amp-sidebar.css (hunk 3 above)
Comment (torch2424) [inline extensions/amp-sidebar/1.0/amp-sidebar.css]: top: 0 breaks the place holder, and doesn't keep it within it's container. Therefore the multiple toolbars will over lap if they all have the same top value.

File: extensions/amp-sidebar/1.0/amp-sidebar.css (hunk 2 above)
Comment (torch2424) [inline extensions/amp-sidebar/1.0/amp-sidebar.css]: Yeah we can do that!

File: extensions/amp-sidebar/1.0/amp-sidebar.css (hunk 1 above)
Comment (camelburrito) [inline extensions/amp-sidebar/1.0/amp-sidebar.css]: Shouldn't the nav only have one UL ?

File: extensions/amp-sidebar/1.0/amp-sidebar.js (hunk 4 above)
Comment (torch2424) [inline extensions/amp-sidebar/1.0/amp-sidebar.js]: I am not re-ording the dom. I am reversing the elements within the array itself. I am doing this because then the toolbars would be added in reverse order. 

I can either do this, or traverse backward with a for(var i = toolbarElements.length) ... loop

File: extensions/amp-sidebar/1.0/amp-sidebar.css (hunk 1 above)
Comment (camelburrito) [inline extensions/amp-sidebar/1.0/amp-sidebar.css]: The way you should ensure it is `.i-amphtml-toolbar > :not(ul:first-child) ` right (or somecss that works?)

File: extensions/amp-sidebar/1.0/toolbar.js (hunk 5 above)
Comment (torch2424) [inline extensions/amp-sidebar/1.0/toolbar.js]: Let's talk offline about this as I cannot find it. 

and this placehodler is used for the elements that will reserve height for our toolbar.

File: extensions/amp-sidebar/1.0/amp-sidebar.css (hunk 3 above)
Comment (camelburrito) [inline extensions/amp-sidebar/1.0/amp-sidebar.css]: That is fine , it is up to the developer to adjust this , but by default it should be 0.

File: extensions/amp-sidebar/1.0/amp-sidebar.js (hunk 4 above)
Comment (camelburrito) [inline extensions/amp-sidebar/1.0/amp-sidebar.js]: Why would it be added in the reverse order?

File: extensions/amp-sidebar/1.0/toolbar.js (hunk 6 above)
Comment (torch2424) [inline extensions/amp-sidebar/1.0/toolbar.js]: The compensating element is within our container that receives display: none; So it would not be shown.

But I can throw this into attempt show before the return.

File: extensions/amp-sidebar/1.0/toolbar.js (hunk 5 above)
Comment (camelburrito) [inline extensions/amp-sidebar/1.0/toolbar.js]: https://www.ampproject.org/docs/guides/responsive/placeholders

Comment (torch2424) [inline extensions/amp-sidebar/1.0/toolbar.js]: Oh that is super cool. Thanks!

File: extensions/amp-sidebar/1.0/amp-sidebar.js (hunk 4 above)
Comment (torch2424) [inline extensions/amp-sidebar/1.0/amp-sidebar.js]: Because we are adding each element as the first child of the body.

So if we have toolbars in our html as:

<body>
first 
second
third
...
<body>

The loop goes as follows when sidebar is applied:

**Iteration 1**

<body>
first 
...
<body>

**Iteration 2**

<body>
second
first 
...
<body>

**Iteration 3**

<body>
third
second
first 
...
<body>

Reversing the array does this in the correct order

Comment (camelburrito) [inline extensions/amp-sidebar/1.0/amp-sidebar.js]: you should not add it as the first child , you will need to add it as the last child so that you dont break existing css. The order of insertion of toolbars should not matter

File: extensions/amp-sidebar/1.0/amp-sidebar.css (hunk 1 above)
Comment (torch2424) [inline extensions/amp-sidebar/1.0/amp-sidebar.css]: That would target the elements that are not the first child of our \<nav toolbar\> . 

Oh wait I was incorrect: https://developer.mozilla.org/en-US/docs/Web/CSS/Child_selectors

It may have been a mistmatch in classes from the AMP start heme, I'll play with it and see what happens

File: extensions/amp-sidebar/1.0/toolbar.js (hunk 5 above)
Comment (torch2424) [inline extensions/amp-sidebar/1.0/toolbar.js]: I added the placeholder, but it does not remove the need for keeping a reference to the placeholder. Since we are resizing it's height as needed.

However, with accordance to your other comment about adding and removing the placeholder dynamically on attemptShow(), this would make sense to move since we would be creating it there, and not in the toolbar build callback. I'll remove this if we decide to go that route instead of hiding the parent container we create.

Comment (camelburrito) [inline extensions/amp-sidebar/1.0/toolbar.js]: ok lets discuss on Monday, also don't call it placeholder :)

File: extensions/amp-sidebar/1.0/amp-sidebar.js (hunk 4 above)
Comment (torch2424) [inline extensions/amp-sidebar/1.0/amp-sidebar.js]: If we do not add the element as the first child then the placeholder solution will not work.

The alternative in that case would to be to add border-top to the body. In that case we would have to maintain the `top:` styling of all our toolbars, as well as getting all of our toolbars to play nicely with one another, or having sidebar manage each toolbar's height to calculate top accordingly. Also, in this case, performing #9983 would be much more difficult

File: extensions/amp-sidebar/1.0/toolbar.js (hunk 5 above)
Comment (torch2424) [inline extensions/amp-sidebar/1.0/toolbar.js]: Sounds good 👍 

I'll rename as I totally forgot to add the placeholder attribute anyways

**Comment by torch2424** [comment]
Split to: https://github.com/torch2424/amphtml/pull/8

**Comment by torch2424** [comment]
Decided the placehodlers would disrupt the `body:first-child` CSS, and we would support only one default toolbar, where it would set the top attribute of the body to the height of the toolbar. Multiple toolbars will conflict with one another, and fight for the space.

**Comment by torch2424** [comment]
The build currently isn't passing, but it appears to be an issue with autosuggest, and not the sidebar component. This is ready for review/merge if able.

File: extensions/amp-sidebar/1.0/toolbar.js [hunk 7]
```diff
@@ -94,12 +103,18 @@ export class Toolbar {
...
     this.toolbarClone_ = this.toolbarDOMElement_.cloneNode(true);
-    this.toolbarClone_.className = TOOLBAR_ELEMENT_CLASS;
+    this.toolbarClone_.classList.add(TOOLBAR_ELEMENT_CLASS);
     this.targetElement_.appendChild(this.toolbarClone_);
     toggle(this.targetElement_, false);
     fragment.appendChild(this.targetElement_);
+
+    // Calculate our body top before returning the fragment
+    this.bodyTop_ =
+      this.win_.getComputedStyle(this.body_, null).getPropertyValue('top');
```
Comment (camelburrito) [inline extensions/amp-sidebar/1.0/toolbar.js]: use the dom.js function

File: extensions/amp-sidebar/1.0/toolbar.js [hunk 8]
```diff
@@ -119,6 +134,17 @@ export class Toolbar {
    * @private
    */
   attemptShow_() {
+
+    // Make room for the toolbar
+    this.vsync_.mutate(() => {
+      if (this.body_) {
+        const toolbarHeight = this.toolbarClone_./*REVIEW*/offsetHeight;
```
Comment (camelburrito) [inline extensions/amp-sidebar/1.0/toolbar.js]: This should be in measure?

File: extensions/amp-sidebar/1.0/toolbar.js [hunk 9]
```diff
@@ -157,6 +183,15 @@ export class Toolbar {
           toggle(element, true);
         });
       }
+
+      // Remove room for our toolbar
+      if (this.body_) {
```
Comment (camelburrito) [inline extensions/amp-sidebar/1.0/toolbar.js]: you should not have to check if body exists.

Comment (torch2424) [inline extensions/amp-sidebar/1.0/toolbar.js]: The check-types linter wont pass without this

File: extensions/amp-sidebar/1.0/toolbar.js [hunk 10]
```diff
@@ -32,6 +35,12 @@ export class Toolbar {
     /** @private {!Object} **/
     this.win_ = win;
 
+    /** @private {Element|null} */
```
Comment (camelburrito) [inline extensions/amp-sidebar/1.0/toolbar.js]: drop the null here and the linter would pass

Comment (torch2424) [inline extensions/amp-sidebar/1.0/toolbar.js]: If you drop the null here you get:

found   : (Element|null)
required: Element
          setStyles(this.body_, {

Comment (torch2424) [inline extensions/amp-sidebar/1.0/toolbar.js]: Error in the linter: ERROR - actual parameter 1 of module$src$style.setStyles does not match formal parameter

File: extensions/amp-sidebar/1.0/toolbar.js [hunk 11]
```diff
@@ -119,6 +132,24 @@ export class Toolbar {
...
   attemptShow_() {
+
+    // Make room for the toolbar
+    this.vsync_.run({
+      measure: state => {
+        if (this.body_ && !this.initialBodyTop_) {
+          this.initialBodyTop_ = computedStyle(this.win_, this.body_)['top'];
+        }
+        state.toolbarHeight = this.toolbarClone_./*REVIEW*/offsetHeight;
+      },
```
Comment (camelburrito) [inline extensions/amp-sidebar/1.0/toolbar.js]: OK
